Index: grpc/src/python/grpcio_observability/setup.py
===================================================================
--- grpc.orig/src/python/grpcio_observability/setup.py
+++ grpc/src/python/grpcio_observability/setup.py
@@ -172,7 +172,6 @@ if EXTRA_ENV_COMPILE_ARGS is None:
         EXTRA_ENV_COMPILE_ARGS += " -fno-wrapv -frtti -fvisibility=hidden"
 
 if EXTRA_ENV_LINK_ARGS is None:
-    EXTRA_ENV_LINK_ARGS = ""
     if "linux" in sys.platform or "darwin" in sys.platform:
         EXTRA_ENV_LINK_ARGS += " -lpthread"
         if check_linker_need_libatomic():
@@ -197,7 +196,7 @@ if BUILD_WITH_STATIC_LIBSTDCXX:
     EXTRA_LINK_ARGS.append("-static-libstdc++")
 
 CC_FILES = [
-    os.path.normpath(cc_file) for cc_file in observability_lib_deps.CC_FILES
+    #os.path.normpath(cc_file) for cc_file in observability_lib_deps.CC_FILES
 ]
 CC_INCLUDES = [
     os.path.normpath(include_dir)
