Description: all Sphinx version build docs
 .
Author: Laszlo Boszormenyi (GCS) <gcs@debian.org>
Bug-Debian: https://bugs.debian.org/955064
Forwarded: no
Last-Update: 2022-11-21

---

--- a/setup.py
+++ b/setup.py
@@ -499,7 +499,7 @@ EXTRAS_REQUIRES = {
 }
 
 SETUP_REQUIRES = INSTALL_REQUIRES + (
-    'Sphinx~=1.8.1',) if ENABLE_DOCUMENTATION_BUILD else ()
+    'Sphinx',) if ENABLE_DOCUMENTATION_BUILD else ()
 
 try:
     import Cython
