Description: all Sphinx version build docs
 .
Author: Laszlo Boszormenyi (GCS) <gcs@debian.org>
Bug-Debian: https://bugs.debian.org/955064
Forwarded: no
Last-Update: 2024-08-11

---

Index: grpc/setup.py
===================================================================
--- grpc.orig/setup.py
+++ grpc/setup.py
@@ -549,7 +549,7 @@ EXTRAS_REQUIRES = {
 }
 
 SETUP_REQUIRES = (
-    INSTALL_REQUIRES + ("Sphinx~=1.8.1",) if ENABLE_DOCUMENTATION_BUILD else ()
+    INSTALL_REQUIRES + ("Sphinx",) if ENABLE_DOCUMENTATION_BUILD else ()
 )
 
 try:
