Description: all Sphinx version build docs
 .
Author: Laszlo Boszormenyi (GCS) <gcs@debian.org>
Bug-Debian: https://bugs.debian.org/955064
Forwarded: no
Last-Update: 2020-06-12

---

--- grpc-1.30.0~pre1.orig/setup.py
+++ grpc-1.30.0~pre1/setup.py
@@ -357,7 +357,7 @@ INSTALL_REQUIRES = (
 )
 
 SETUP_REQUIRES = INSTALL_REQUIRES + (
-    'Sphinx~=1.8.1',
+    'Sphinx',
     'six>=1.10',
 ) if ENABLE_DOCUMENTATION_BUILD else ()
 
